### ISSUE TITLE
Sudden segfault

### ISSUE THREAD

**Issue by Kenny690** [e2guardian/e2guardian#213]
IDK what's happened, but this is what I get in syslog:

`Apr 14 15:05:46 osiris kernel: [ 2579.294950] e2guardian[1831]: segfault at 478d178 ip 00007f048e2ffef3 sp 00007f02ecde4238 error 4 in libc-2.24.so[7f048e27c000+195000]`

Last notes in e2guardian:

```
2017.4.14 15:05:42 - 10.20.30.84 https://s.youtube.com/api/stats/qoe?event=streamingstats&fmt=278&afmt=250&cpn=6r_DenzFMQ764Gkw&ei=7p7wWPiVLIyCY7PJlfAJ&el=detailpage&docid=1VjPmTlEH0U&ns=yt&fexp=3300118%2C3300132%2C3300164%2C3313275%2C3313321%2C9417828%2C9422596%2C9428398%2C9431012%2C9434046%2C9434289%2C9434797%2C9446054%2C9446364%2C9449243%2C9450707%2C9453078%2C9453897%2C9456640%2C9457141%2C9457166%2C9457439%2C9458576%2C9460160%2C9462229%2C9462397%2C9463154%2C9463496%2C9463594%2C9463882%2C9463965%2C9464485%2C9465353%2C9465533%2C9465616%2C9466783%2C9466783%2C9466793%2C9466795%2C9467030%2C9467533%2C9467883%2C9468166%2C9468290%2C9468350%2C9468482%2C9469224%2C9469637%2C9470265%2C9471390%2C9471471&cl=152966617&seq=2&c=WEB&cver=1.20170412&cplayer=UNIPLAYER&cbr=Chrome&cbrver=57.0.2987.133&cos=Windows&cosver=6.1&cmt=0.339:0.000,2.127:0.000&vps=0.339:B,2.127:B&ctmp=msst:t.0.346;vbr.20748.625;abr.11560.5;bwe.440.0767018216682&vfs=2.127:278:247:278:r&view=2.127:1366:768&bwe=2.127:440&bh=2.127:0.000&bat=2.127:1:1 - GET 0 0 - 1 204 text/html - default - - - - -
2017.4.14 15:05:45 - 10.20.30.84 https://www.youtube.com/annotations_invideo?cap_hist=1&video_id=1VjPmTlEH0U&client=1&ei=7p7wWPiVLIyCY7PJlfAJ *SCANNED* - POST 2786 0 - 1 200 text/xml - default - - - - -
2017.4.14 15:05:45 - 10.20.30.84 https://www.youtube.com/player_204?rt=5.224&allowed=1_2%2C1_2_1%2C2_2_1%2C2_2%2C2_3&ex_ad=46%3D14_14%3B59%3D14_14%3B77%3D1_1%2C1_3%2C2_1%2C2_3%3B89%3D2_2%2C15_2%3B94%3D15_2%2C15_2_1%2C15_2_4%2C17_2%2C17_2_1&fs=1&volume=100&mos=0&fmt=278&afmt=250&plid=AAVNHZUgyChWX4QI&ei=9J7wWNCAFde7YMyLteAM&cpn=6r_DenzFMQ764Gkw&v=1VjPmTlEH0U&ns=yt&html5=1&el=detailpage&feature=search-shelf&l-class=3&xble=0&a-ch=UCs_uv3QyUIQjBoL1Ij5BdlQ&link-id=UCs_uv3QyUIQjBoL1Ij5BdlQ&a-v=1VjPmTlEH0U&a-type=10&a-id=e1892cbc-6970-2db1-81cb-485b395e7705&p-time=0.00&iv-event=2&event=iv&ux=1 - GET 0 0 - 1 204 text/html - default - - - - -
2017.4.14 15:05:45 - 10.20.30.84 https://www.youtube.com/get_endscreen?v=1VjPmTlEH0U&ei=7p7wWPiVLIyCY7PJlfAJ&client=1 *SCANNED* - POST 8533 0 - 1 200 text/plain - default - - - - -

```

**Comment by fredbcode** [comment]
With strech ? compiled or from my package ?

**Comment by Kenny690** [comment]
Yes.
I've got second few seconds ago. Seems like I know the reason, but not clearly. The last thing I did - is rewrote iptables rules.

**Comment by Kenny690** [comment]
Here's them:

```
#!/bin/bash
export IPT="iptables"

# внешка
export WAN=eth0
export WAN_IP=10.20.30.130

# локалка

export LAN=eth2
export LAN_IPRNG=10.20.30.0/24

# Блочим тех, кто не соответствует правилам ниже

$IPT -P INPUT DROP
$IPT -P OUTPUT DROP
$IPT -P FORWARD DROP

# Разрешаем трафик локалхоста и локалки

$IPT -A INPUT -i lo -j ACCEPT
$IPT -A INPUT -i $LAN -j ACCEPT
$IPT -A OUTPUT -o lo -j ACCEPT
$IPT -A OUTPUT -o $LAN -j ACCEPT

# Даём ICMP запросы

$IPT -A INPUT -p icmp --icmp-type echo-reply -j ACCEPT
$IPT -A INPUT -p icmp --icmp-type destination-unreachable -j ACCEPT
$IPT -A INPUT -p icmp --icmp-type time-exceeded -j ACCEPT
$IPT -A INPUT -p icmp --icmp-type echo-request -j ACCEPT
$IPT -A FORWARD -p icmp --icmp-type echo-request -j ACCEPT

# Разрешаем lm для админа
$IPT -A FORWARD -p tcp -s 10.20.30.50 --dport 5650 -j ACCEPT

$IPT -A OUTPUT -o $WAN -j ACCEPT

# НЕ УБИРАТЬ КОММЕНТ БЕЗ НАДОБНОСТИ!

#$IPT -A INPUT -i $WAN -j ACCEPT

# Убираем повторную проверку

$IPT -A INPUT -p all -m state --state ESTABLISHED,RELATED -j ACCEPT
$IPT -A OUTPUT -p all -m state --state ESTABLISHED,RELATED -j ACCEPT
$IPT -A FORWARD -p all -m state --state ESTABLISHED,RELATED -j ACCEPT

# Защита от пакетов без статуса

$IPT -A INPUT -m state --state INVALID -j DROP
$IPT -A FORWARD -m state --state INVALID -j DROP

# Защита от нулевых пакетов

$IPT -A INPUT -p tcp --tcp-flags ALL NONE -j DROP

# Убираем syn-flood атаки

$IPT -A INPUT -p tcp ! --syn -m state --state NEW -j DROP
$IPT -A OUTPUT -p tcp ! --syn -m state --state NEW -j DROP

# Блокируем IP

#$IPT -A INPUT -s 8.8.8.8 -j REJECT

# Разрешаем интернет

$IPT -A FORWARD -i $LAN -o $WAN -j ACCEPT

# Запрещаем к нам заходить извне

$IPT -A FORWARD -i $WAN -o $LAN -j REJECT
$IPT -t nat -A POSTROUTING -o $WAN -s $LAN_IPRNG -j MASQUERADE

$IPT -A INPUT -i $WAN -p udp --dport 53 -j ACCEPT

# Логирование

$IPT -N block_in
$IPT -N block_out
$IPT -N block_fwd

$IPT -A INPUT -j block_in
$IPT -A OUTPUT -j block_out
$IPT -A FORWARD -j block_fwd

$IPT -A block_in -j LOG --log-level info --log-prefix "--IN--BLOCK"
$IPT -A block_in -j DROP
$IPT -A block_out -j LOG --log-level info --log-prefix "--OUT--BLOCK"
$IPT -A block_out -j DROP
$IPT -A block_fwd -j LOG --log-level info --log-prefix "--FWD--BLOCK"
$IPT -A block_fwd -j DROP
```

**Comment by Kenny690** [comment]
And yes, I have logging, but no messages appears before segfault.

**Comment by fredbcode** [comment]
I guess you should compile, the package was made for stable OS (Jessie)

**Comment by Kenny690** [comment]
I'll try to figure out why this is happening. Right now 20 minutes no segfaults. I think it's some website causing it.

**Comment by fredbcode** [comment]
Ok thanks, I guess there is a problem with some libraries  segfault at 478d178 ip 00007f048e2ffef3 sp 00007f02ecde4238 error 4 in libc-2.24.so

My binary was made for previous debian version

**Comment by fredbcode** [comment]
The issue about systemd is fixed now ?

**Comment by Kenny690** [comment]
Found it! Sometimes I'm getting segfault when I'm trying to get access to mail.ru. The loading takes me a long time and sometimes it's not successful. Can you check it?

**Comment by fredbcode** [comment]
Sorry I'm out of office now, I will next week .
An account is needed ?

Meanwhille maybe you can make a test with a stable release, Ubuntu or debian ?

Le 17 avril 2017 14:00:29 GMT+02:00, Kenny690 <notifications@github.com> a écrit :
>Found it! Sometimes I'm getting segfault when I'm trying to get access
>to mail.ru. The loading takes me a long time and sometimes it's not
>successful. Can you check it? 
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-294472172

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by Kenny690** [comment]
No, account is not needed. I'll try to find some free PC to test it. Don't have it right now thou.

**Comment by ktullanux** [comment]
+1 same problem with mail.ru 
ubuntu 16.04.2 LTS amd64

**Comment by fredbcode** [comment]
@ktullanux with 4.1 ?

**Comment by ktullanux** [comment]
@fredbcode yep! build 4.1 at 13.04.2017 13:43 (Moscow time) with 
`Built with: '--prefix=/usr' '--enable-clamd=yes' '--with-proxyuser=e2guardian' '--with-proxygroup=e2guardian' '--sysconfdir=/etc' '--localstatedir=/var' '--enable-icap=yes' '--enable-commandline=yes' '--enable-email=yes' '--enable-ntlm=yes' '--enable-trickledm=yes' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--enable-pcre=yes' '--enable-sslmitm=yes'`

**Comment by fredbcode** [comment]
Ok thanks it works with SSLMITM off ?

**Comment by ktullanux** [comment]
works fine, if sslmitm = off :)

**Comment by fredbcode** [comment]
I did try to reproduce, home page OK, but did not know where to go next to test as I do not know Russian

Le 17 avril 2017 18:29:47 GMT+02:00, Artem Lukyanov <notifications@github.com> a écrit :
>works fine, if sslmitm = off :)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-294521912

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by Kenny690** [comment]
You have no delays before open? The homepage has this problem too. e.mail.ru acting weird. In my case it sometimes shows itself differently.

**Comment by fredbcode** [comment]
Please, can you take a look at your CPU usage when you hit the page ?

Le 18 avril 2017 10:09:23 GMT+02:00, Kenny690 <notifications@github.com> a écrit :
>You have no delays before open? The homepage has this problem too.
>e.mail.ru acting weird. In my case it sometimes shows itself
>differently. 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-294722943

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by fredbcode** [comment]
And do you have something in syslog and e2 log ?

Le 18 avril 2017 10:09:23 GMT+02:00, Kenny690 <notifications@github.com> a écrit :
>You have no delays before open? The homepage has this problem too.
>e.mail.ru acting weird. In my case it sometimes shows itself
>differently. 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-294722943

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by Kenny690** [comment]
I turned off everyone from e2guardian, just putted some testing machine. And I turned stopwatch just to test the time.
There is no e2guardian process at all after I entered website's domain and press enter button. I need to wait for 1 min 5 sec. Then in top appears e2guardian process that eats ~70% of CPU. Then webstie loads normally.

access.log:

```
2017.4.18 13:28:57 - 10.20.30.50 https://mail.ru - GET 262144 -365 - 1 200 text/html - default - - - - -
2017.4.18 13:29:09 - 10.20.30.50 https://mail.ru - GET 262144 -365 - 1 200 text/html - default - - - - -
2017.4.18 13:29:10 - 10.20.30.50 https://top-fwz1.mail.ru/counter?js=13;id=110605;u=https%3A//mail.ru/;st=1492503934627;title=Mail.Ru%3A%20%D0%BF%D0%BE%D1%87%D1%82%D0%B0%2C%20%D0%BF%D0%BE%D0%B8%D1%81%D0%BA%20%D0%B2%20%D0%B8%D0%BD%D1%82%D0%B5%D1%80%D0%BD%D0%B5%D1%82%D0%B5%2C%20%D0%BD%D0%BE%D0%B2%D0%BE%D1%81%D1%82%D0%B8%2C%20%D0%B8%D0%B3%D1%80%D1%8B;s=1920*1080;vp=1202*887;touch=0;hds=1;flash=25.0;sid=9b05aeac07e61392;ver=60;_=0.8409936214613778 *DENIED*  URL: top-fwz1.mail.ru/counter GET 0 0 - 1 403 image/gif - default - - - - -
2017.4.18 13:29:10 - 10.20.30.50 https://bar.love.mail.ru/jsonp/bar?rnd=1492503934786 *DENIED*  : love.mail.ru GET 0 0 - 1 403 application/javascript - default - - - - -
2017.4.18 13:29:10 - 10.20.30.50 https://counter.yadro.ru/hit;mail-splash/pc?r;s1920*1080*24;uhttps%3A//mail.ru/;0.4864776872779337 *DENIED*  : yadro.ru GET 0 0 - 1 403 image/gif - default - - - - -
2017.4.18 13:29:10 - 10.20.30.50 https://ok.ru/mapi?query=%7B%22cmd%22%3A%22getCounters%22%7D&callback=__PHJSONPCallback_0&rnd=1492503934787 *DENIED*  : ok.ru GET 0 0 - 1 403 application/x-javascript - default - - - - -
2017.4.18 13:29:10 - 10.20.30.50 https://r.mradx.net/h5/BA/9673CABE/8724CC0C24.html?data=CInsrwyIAQfgAQEQwqarTx1WzvVYIAUoAaAB8cCKyQi4AQuQA_HAislowgIMyzmUC74LthClBMEL&link1=252603714 - GET 1547 0 - 1 200 text/html - default - - - - -
```

syslog:

```
Apr 18 13:29:04 osiris e2guardian[12943]:  -connection handler caught an exception std::bad_alloc
```

That's all.

**Comment by fredbcode** [comment]
Can you check with rules reduced, just SSLMITM and light blacklists/regex ?

Le 18 avril 2017 10:32:55 GMT+02:00, Kenny690 <notifications@github.com> a écrit :
>I turned off everyone from e2guardian, just putted some testing
>machine. And I turned stopwatch just to test the time.
>There is no e2guardian process at all after I entered website's domain
>and press enter button. I need to wait for 1 min 5 sec. Then in top
>appears e2guardian process that eats ~70% of CPU. Then webstie loads
>normally.
>
>access.log:
>
>```
>2017.4.18 13:28:57 - 10.20.30.50 https://mail.ru - GET 262144 -365 - 1
>200 text/html - default - - - - -
>2017.4.18 13:29:09 - 10.20.30.50 https://mail.ru - GET 262144 -365 - 1
>200 text/html - default - - - - -
>2017.4.18 13:29:10 - 10.20.30.50
>https://top-fwz1.mail.ru/counter?js=13;id=110605;u=https%3A//mail.ru/;st=1492503934627;title=Mail.Ru%3A%20%D0%BF%D0%BE%D1%87%D1%82%D0%B0%2C%20%D0%BF%D0%BE%D0%B8%D1%81%D0%BA%20%D0%B2%20%D0%B8%D0%BD%D1%82%D0%B5%D1%80%D0%BD%D0%B5%D1%82%D0%B5%2C%20%D0%BD%D0%BE%D0%B2%D0%BE%D1%81%D1%82%D0%B8%2C%20%D0%B8%D0%B3%D1%80%D1%8B;s=1920*1080;vp=1202*887;touch=0;hds=1;flash=25.0;sid=9b05aeac07e61392;ver=60;_=0.8409936214613778
>*DENIED*  URL: top-fwz1.mail.ru/counter GET 0 0 - 1 403 image/gif -
>default - - - - -
>2017.4.18 13:29:10 - 10.20.30.50
>https://bar.love.mail.ru/jsonp/bar?rnd=1492503934786 *DENIED*  :
>love.mail.ru GET 0 0 - 1 403 application/javascript - default - - - - -
>2017.4.18 13:29:10 - 10.20.30.50
>https://counter.yadro.ru/hit;mail-splash/pc?r;s1920*1080*24;uhttps%3A//mail.ru/;0.4864776872779337
>*DENIED*  : yadro.ru GET 0 0 - 1 403 image/gif - default - - - - -
>2017.4.18 13:29:10 - 10.20.30.50
>https://ok.ru/mapi?query=%7B%22cmd%22%3A%22getCounters%22%7D&callback=__PHJSONPCallback_0&rnd=1492503934787
>*DENIED*  : ok.ru GET 0 0 - 1 403 application/x-javascript - default -
>- - - -
>2017.4.18 13:29:10 - 10.20.30.50
>https://r.mradx.net/h5/BA/9673CABE/8724CC0C24.html?data=CInsrwyIAQfgAQEQwqarTx1WzvVYIAUoAaAB8cCKyQi4AQuQA_HAislowgIMyzmUC74LthClBMEL&link1=252603714
>- GET 1547 0 - 1 200 text/html - default - - - - -
>```
>
>syslog:
>
>```
>Apr 18 13:29:04 osiris e2guardian[12943]:  -connection handler caught
>an exception std::bad_alloc
>```
>
>That's all.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-294729350

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by Kenny690** [comment]
I did it yesterday, but forgot to write about it. I tried to remove everything from banndesitelist and bannedurllist. Nothing changed. 
The only thing that helped - I added IP of a client into the exceptioniplist.

**Comment by fredbcode** [comment]
No segfault here with https://mail.ru
I'm testing with your configs files on Debian Jessie 64 bits
I have a little test machine (only 2 cores, 2Go of RAM), sometimes I'm using 150% of CPU but it works well

I can access mail.ru after more or less 5/10 seconds

**Comment by Kenny690** [comment]
I tried to completely remove e2guardian, reintall it and with default settings, just with turned on sslmitm and enablessl I typed in my browser "mail.ru". Still has that problem.
I have one suggestion. Did you try my certificates when you test it?

**Comment by Kenny690** [comment]
Hm... Found something new. I tried to set proxyexchange = 5 and I've got mail.ru in 5 sec. Maybe the problem is in proxy. I will investigate it.

**Comment by fredbcode** [comment]
Ok, Yes I tested with your files, certificats included

Le 22 avril 2017 06:52:17 GMT+02:00, Kenny690 <notifications@github.com> a écrit :
>Hm... Found something new. I tried to set proxyexchange = 5 and I've
>got mail.ru in 5 sec. Maybe the problem is in proxy. I will investigate
>it.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-296347545

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by philipianpearce** [comment]
I have pushed a temporary branch called 'header' which contains a trial fix for a similar issue we are seeing.      
Would you try it and see if it fixes your issue?
Please let me know the result.

Thanks

**Comment by fredbcode** [comment]
Philip you have pushed to the dev branch not v4.1, I guess it should be in v4.1 ?

FYI, I made some changes to xforwarded_for especially with NTLM, I wonder if there is also some relations with SSLMITM (eg some headers are twice ...) 

Eg: I was stuck with https://www.sciencesetavenir.fr + NTLM + SSLMITM and now it works 

I'm not happy with identification, especially NTLM, I tried to remove ugly hacks like adding xforwaderd_for header in this part of code (still one) but it needs deeper review ...

@Kenny690 I guess v4.1 should works better for you now, especially logs

**Comment by fredbcode** [comment]
> https://www.sciencesetavenir.fr + NTLM + SSLMITM
No just related with SSLMITM, sometime if I try a link (bottom of the page) like "En images" "L'essentiel santé" I have a blank page

**Comment by dkwiebe** [comment]
We've been running e2guardian 3.5 for quite some time and it's been working well.  Last night I upgraded to 4.1 and we're getting segfaults as well.

10442.838778] TCP: request_sock_TCP: Possible SYN flooding on port 8085. Sending cookies.  Check SNMP counters.
[10508.371210] e2guardian[21901]: segfault at fffffffffffffff8 ip 00007f1f93aa62e4 sp 00007f1ee0fd5940 error 5 in libstdc++.so.6.0.20[7f1f939e6000+ec000]
[10986.392811] e2guardian[22940]: segfault at fffffffffffffff8 ip 00007f9c945db2e4 sp 00007f9b61ed7940 error 5 in libstdc++.so.6.0.20[7f9c9451b000+ec000]
[11165.351337] e2guardian[23581]: segfault at fffffffffffffff8 ip 00007fc39a2f92e4 sp 00007fc2b17ee940 error 5 in libstdc++.so.6.0.20[7fc39a239000+ec000]
[11504.486279] e2guardian[24160]: segfault at fffffffffffffff8 ip 00007f41f2cd62e4 sp 00007f40ebf2b940 error 5 in libstdc++.so.6.0.20[7f41f2c16000+ec000]
[12041.489759] e2guardian[25632]: segfault at fffffffffffffff8 ip 00007f11b0dd22e4 sp 00007f108fefb940 error 5 in libstdc++.so.6.0.20[7f11b0d12000+ec000]
[12418.157754] e2guardian[26081]: segfault at 72cb2d8 ip 00007f6127f1f04e sp 00007f606e7c7238 error 4 in libc-2.19.so[7f6127e8d000+1a1000]
[12755.919808] e2guardian[26771]: segfault at fffffffffffffff8 ip 00007f1c268c32e4 sp 00007f1b8dff7940 error 5 in libstdc++.so.6.0.20[7f1c26803000+ec000]


It doesn't seem to be related to any specific ssl sites.  Or at least I haven't been able to pin it down to any.  I'd be glad to pay for your time to take a look at it.

**Comment by philipianpearce** [comment]
See below 

Regards 

Philip 

----- Original Message -----

> From: "Fredb" <notifications@github.com>
> To: "e2guardian/e2guardian" <e2guardian@noreply.github.com>
> Cc: "Philip Pearce" <philip.pearce@e2bn.org>, "Comment"
> <comment@noreply.github.com>
> Sent: Tuesday, 25 April, 2017 5:20:14 PM
> Subject: Re: [e2guardian/e2guardian] Sudden segfault (#213)

> Philip you have pushed to the dev branch not v4.1, I guess it should
> be in v4.1 ?
No, I pushed temporary branch 'header' as a fork from v4.1 - I just checked on github and this is correct. Once fix is complete and proved I will merge back into v4.1 and delete 'header' branch. 

The push to v4develop is work in progress on next version, which will compile but likely not run. 

> FYI, I made some changes to xforwarded_for especially with NTLM, I
> wonder if there is also some relations with SSLMITM (eg some headers
> are twice ...)
This should not happen (2 headers) with or without sslmitm unless X-Forwarded-For was already set in the request. 

When in a MITM session (i.e. when ismitm is set) X-Forwarded-For must not be added as we are talking directly to the target server (via tunnel through squid). Squid will never see the header and target server may get upset! 

NTLM plugin does it's own dance with squid to get auth before squid returns good http response (or tunneling with CONNECT), so X-Forwarded-For should be added to these requests so that squid logs right IP. 

> Eg: I was stuck with https://www.sciencesetavenir.fr + NTLM + SSLMITM
> and now it works
> I'm not happy with identification, especially NTLM, I tried to remove
> ugly hacks like adding xforwaderd_for header in this part of code
> (still one) but it needs deeper review ...
> @Kenny690 I guess v4.1 should works better for you now, especially
> logs
I will take a look, when I get a chance.

**Comment by philipianpearce** [comment]
@Darren 

Can you try the test 'header' branch and see if this cures the problem? 

Also worth checking if any of your users are running proxy avoidance software, probes or viral activity, especially as you have a SYN flood report. 

Regards 

Philip 

----- Original Message -----

**Comment by dkwiebe** [comment]
I installed the header branch this morning to test but that hasn't seemed to make a difference.  We're running squid 3.1.20 in front of this one so the SYN flood would have hit squid.

**Comment by fredbcode** [comment]
Yes Philip, I guess it's good now with SSLMITM ( or without ) and NTLM. 
There were some duplicates values like xforwaded_for and keep-alive, I cleaned and the result seem better with tcpdump now.

What I mean about SSLMITM, it's maybe there is a link with some strange issues and NTLM unexpected headers (but maybe not)

Sorry for the mistake about the branch

**Comment by fredbcode** [comment]
@dkwiebe which os ? Compilation ? Package ?
It works well without SSLMITM ?

**Comment by dkwiebe** [comment]
Debian 8 compiled from source.  We haven't tested without SSLMITM

e2guardian 4.1.0

Built with:  '--enable-sslmitm' '--prefix=/usr' '--sysconfdir=/etc' '--enable-clamd=yes' '--with-proxyuser=e2guardian' '--with-proxygroup=e2guardian' '-localstatedir=/var' '--enable-icap=yes' '--enable-commandline=yes' '--enable-email=yes' '--enable-ntlm=yes' '--enable-trickledm=yes' '--with-piddir=/var/run'

I'm in the process of preparing another test box.

**Comment by fredbcode** [comment]
Have you tested with our debian package ?

**Comment by dkwiebe** [comment]
No, I hadn't thought of that.  I'll do more testing with that and get back to you.

**Comment by Kenny690** [comment]
With "headers" I have the same issue. Can't test segfaults, cause they're very sudden, but mail.ru still taking too long to load.

**Comment by fredbcode** [comment]
@Kenny690 Can you try v4.1 latest pushes, at least for bug #212 
Thanks

**Comment by fredbcode** [comment]
Oups I forgot, and make a try with NTLM disabled, just in case (but it works well here with your configuration)

**Comment by Kenny690** [comment]
With 4.1 and turned off NTLM auth nothing's changed. Except of squid is logging client's IP right.

**Comment by Kenny690** [comment]
There is some differences. I have a lot of this in logs. Like really a lot:

```
Apr 26 12:15:40 osiris e2guardian[10600]: NTLM - Invalid message of length 0, message was:
Apr 26 12:15:40 osiris e2guardian[10600]: Auth plugin returned error code: -3
```

**Comment by fredbcode** [comment]
Ok there is no relation please report this to #212

**Comment by Kenny690** [comment]
Woah-waoh-woah! What have you done? Mail.ru is now opening instantly. I'll check everything for segfaults. Thank you!

**Comment by Kenny690** [comment]
Oh, I know what. I have enablessl=on and sslmitm=on, but I don't have ssl decoding:
```
2017.4.26 13:27:51 - 10.20.30.80 https://e.mail.ru:443 - CONNECT 3670 0 - 1 0 - - default - - - - -
2017.4.26 13:27:51 - 10.20.30.80 https://e.mail.ru:443 - CONNECT 3670 0 - 1 0 - - default - - - - -
2017.4.26 13:27:51 - 10.20.30.80 https://gstat.imgsmail.ru:443 - CONNECT 3702 0 - 1 0 - - default - - - - -
2017.4.26 13:27:53 - 10.20.30.217 https://ok.ru:443 - CONNECT 3655 0 - 1 0 - - default - - - - -
2017.4.26 13:27:53 - 10.20.30.217 https://bar.love.mail.ru:443 - CONNECT 3699 0 - 1 0 - - default - - - - -
2017.4.26 13:27:58 - 10.20.30.215 https://clients4.google.com:443 - CONNECT 3705 0 - 1 0 - - default - - - - -
2017.4.26 13:28:12 - 10.20.30.100 https://bar.love.mail.ru:443 - CONNECT 3699 0 - 1 0 - - default - - - - -
2017.4.26 13:28:13 - 10.20.30.142 https://bar.love.mail.ru:443 - CONNECT 3688 0 - 1 0 - - default - - - - -
2017.4.26 13:28:14 - 10.20.30.100 https://ok.ru:443 - CONNECT 3655 0 - 1 0 - - default - - - - -
2017.4.26 13:28:15 - 10.20.30.142 https://ok.ru:443 - CONNECT 3644 0 - 1 0 - - default - - - - -
2017.4.26 13:27:51 - 10.20.30.80 https://e.mail.ru:443 - CONNECT 3670 0 - 1 0 - - default - - - - -
2017.4.26 13:28:26 - 10.20.30.100 https://e.mail.ru:443 - CONNECT 3671 0 - 1 0 - - default - - - - -
```

**Comment by fredbcode** [comment]
?? I'm pretty sure that this is not related with my latest changes (only in ntlm.cpp)
What do you have when you start with e2guardian -N

**Comment by Kenny690** [comment]
Auth plugin returned error code: -3
That's all.

**Comment by fredbcode** [comment]
And you have the right certificate in your browser ? --enable-sslmitm=yes is not missing ?

**Comment by Kenny690** [comment]
root@Osiris:~/e2guardian# e2guardian -v
e2guardian 4.1.0

Built with:  '--prefix=/usr' '--enable-clamd=yes' '--with-proxyuser=e2guardian' '--with-proxygroup=e2guardian' '--sysconfdir=/etc' '--localstatedir=/var' '--enable-icap=yes' '--enable-commandline=yes' '--enable-email=yes' '--enable-ntlm=yes' '--enable-trickledm=yes' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' 'CXXFLAGS=-g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-z,relro' 'CPPFLAGS=-D_FORTIFY_SOURCE=2' 'CFLAGS=-g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security' '--enable-pcre=yes' '--enable-locallists=yes' '--enable-sslmitm=yes'
Yep.

**Comment by fredbcode** [comment]
I'm surfing right now with your certificate "School 30" on mail.ru, to be sure can you make a try with a comment before authplugin= yourpath/proxy-ntlm.conf

**Comment by Kenny690** [comment]
The same without ntlm. But I have a lot of this when I'm trying to get at mail.ru

```
Apr 27 14:14:32 osiris e2guardian[27193]: 51353 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51354 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51355 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51356 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51357 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51358 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51359 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51361 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51363 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51362 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51360 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51364 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51385 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51386 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51365 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51366 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51367 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51371 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51368 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51369 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51372 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51370 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51373 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51379 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51374 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51378 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51377 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51375 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51376 No header recd from client at 10.20.30.128 - errno: 0
Apr 27 14:14:32 osiris e2guardian[27193]: 51380 No header recd from client at 10.20.30.128 - errno: 0
```

**Comment by philipianpearce** [comment]
If you use firebug you will likely see that the site is using javascript calls using websocket which are failing, as they are not http - hence "No (http) header recd from client". 

In normal use "No header recd from client" can be ignored. If logconnectionhandlingerrors is off then the message will not appear. 

Regards 

Philip 

----- Original Message -----

**Comment by Kenny690** [comment]
One day passed no segfaults yet. But there is still problems with opening some websites yet.

**Comment by 1nfrag** [comment]
I have random segfaults similar to yours Kenny. I've re-setup a clean install of debian with the same result using the debian package. I havent tried a source compile yet for segfault fix.
I have compiled from source to check if it fixed loading certain https sites. It doesnt fix them. A clean install and sites like: Americanexpress.com.au dont work. Even hotmail.com struggles to load or anz.com.au. (Sites like https://google.com.au or https://bing.com) work fine. If I turn off mitm the sites work fine. The sites are completely dead but take a long time and then only a garbled version of the site appears. 

Browser console shows: Failed to load resource: net::ERR_EMPTY_RESPONSE for what I can tell are linked sites that its trying to load information from.

Sorry I dont mean to hijack the thread but from what I read it seems very similar.

**Comment by fredbcode** [comment]
Which OS ? Did you see high CPU usage ? 

Do you have something in E2's log ? Squid ?

**Comment by philipianpearce** [comment]
@Infrag

I've had a look at the Amex site and can replicate the slowness.    It is a complex javascript site which is issuing inline calls to advert sites which are being blocked.   However e2g is returning text/html and not javascript.     This is not an issue specific to sslmitm, it is just that when you have mitm off these urls are not being inspected and are let through.

I have found a way to fix the symptoms - switch off advert blocking.

It does not seem to have any relation to crashes - I can replicate here without a crash and we have many systems on v4.1 in large-scale production and we are not having any crashes at all.

I will open a issue on this and we will look at in next version.

**Comment by fredbcode** [comment]
>  However e2g is returning text/html and not javascript. 

How how it can be ? I mean we have no problem with javascript usually ?

**Comment by fredbcode** [comment]
Perhaps we can do as flash, return a fake javascript ?

**Comment by philipianpearce** [comment]
Problems are to do with the calling javascript - javascript is calling url (without .js extension) and e2g is detecting an advert and returning a text/html advert block page. Calling javascript is not even checking the return type, but is executing it as js. Get firebug error and hangs! 

Where the calling javascript does check the return type, before executing js, all is OK so no problem on most sites. 

Regards 

Philip 

----- Original Message -----

**Comment by fredbcode** [comment]
It's just a matter of extension ? But perphaps we can detect the type and switch to another js or maybe an empty file ...
In this case the browser should have a wrong JS but not an unexpected file type ?

Le 2 mai 2017 18:17:46 GMT+02:00, Philip Pearce <notifications@github.com> a écrit :
>Problems are to do with the calling javascript - javascript is calling
>url (without .js extension) and e2g is detecting an advert and
>returning a text/html advert block page. Calling javascript is not even
>checking the return type, but is executing it as js. Get firebug error
>and hangs! 
>
>Where the calling javascript does check the return type, before
>executing js, all is OK so no problem on most sites. 
>
>Regards 
>
>Philip 
>
>----- Original Message -----
>
>> From: "Fredb" <notifications@github.com>
>> To: "e2guardian/e2guardian" <e2guardian@noreply.github.com>
>> Cc: "Philip Pearce" <philip.pearce@e2bn.org>, "Comment"
>> <comment@noreply.github.com>
>> Sent: Tuesday, 2 May, 2017 4:07:40 PM
>> Subject: Re: [e2guardian/e2guardian] Sudden segfault (#213)
>
>> > However e2g is returning text/html and not javascript.
>> 
>> How how it can be ? I mean we have no problem with javascript usually
>> ?
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub , or mute the thread
>> .
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-298684453

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by fredbcode** [comment]
@kenny690 @Infrag 
Can you try without textmimetypes= 
If you have this option comment/remove it and restart.

Tell us if you see something, I guess not but ...

**Comment by Kenny690** [comment]
I don't have this option.

**Comment by fredbcode** [comment]
Ok Kenny please try with this line commented 
In bannesitelist 
/etc/e2guardian/lists/bannedsitelist:# .Include</etc/e2guardian/lists/blacklists/ads/domains>

@philipianpearce With american express I can also see some "HTTP/1.1 407 Proxy Authentication Required" do you have this ?

**Comment by fredbcode** [comment]
I always have 407 returns code with:
https://www.americanexpress.com/australia/rwd/heroes/40066/hero.html?type=pzn_error 
( not directly , but with https://americanexpress.com.au )

I'm using digest authentication, do you have the same thing ?

**Comment by 1nfrag** [comment]
I dont use authentication currently on the sites I've upgraded to 4.1
I'll try turning off AD blocking tomorrow and see if this fixes things.
Thanks for looking into it and the prompt replies!

**Comment by fredbcode** [comment]
I guess I found something 

Same problem with mail.ru and basic ident, Philip I think there is a bug here 
No problem at all with enablessl=off and HTTP sites

**Comment by fredbcode** [comment]
@1nfrag This is unexpected behaviour do you see some lines with 407 (e2's log) without any identification ?
With loglevel 3 ?

**Comment by philipianpearce** [comment]
@fred 

Yes, it is this similar type of problem as with mail.ru, but not exactly the same. 

I do not not use any pass-through auth and so do not get any 407's. I can still replicate the problem with Ad blocking. But the 407 may also be explained by the browser not processing the response as normal. These sites are using javascript to make these requests and these are not being processed by the standard http browser engine, but by the browser javascript engine which is not checking the mime type returned. This should be able to be seen in firebug. 

Or the 407 lines may just be a result of the changes you have made to the logging? There are lots of different sites called by the javascript on the amex site, including 3 advert sites. Each of these sites will result in a new connection and 407's being returned from squid. 

Do you get the 407's if you go directly to squid? 

The majority of https sites that use javascript (inc google etc) work good with sslmitm. It is just a few sites with this problem and these sites happen to be https. 

It is a bug, but a bug in the javascript of these sites which we have to try and work round without breaking all the other sites which currently work ok. 

I have some ideas than may work with the current release which I will try later today, which may solve the advert issue. 

Re mail.ru - does this work OK without indent? 

What is the symptom when indent is on? firebug trace? 

Regards 

Philip 

----- Original Message -----

**Comment by fredbcode** [comment]
> @fred
> 
> Yes, it is this similar type of problem as with mail.ru, but not
> exactly the same.
> 
> I do not not use any pass-through auth and so do not get any 407's. I
> can still replicate the problem with Ad blocking. But the 407 may
> also be explained by the browser not processing the response as
> normal. These sites are using javascript to make these requests and
> these are not being processed by the standard http browser engine,
> but by the browser javascript engine which is not checking the mime
> type returned. This should be able to be seen in firebug.
> 

And, maybe the javascript engine just doesn't replay identification, but even in this case the website should doesn't answer to a unexpected HTTP request  


> Or the 407 lines may just be a result of the changes you have made to
> the logging? There are lots of different sites called by the
> javascript on the amex site, including 3 advert sites. Each of these
> sites will result in a new connection and 407's being returned from
> squid.

Maybe, but I don't think, the changes I made just get the 407 
But you are right to be sure I will retry with 3.5.1

> 
> Do you get the 407's if you go directly to squid?
> 

Not at all 

> The majority of https sites that use javascript (inc google etc) work
> good with sslmitm. It is just a few sites with this problem and
> these sites happen to be https.
> 
> It is a bug, but a bug in the javascript of these sites which we have
> to try and work round without breaking all the other sites which
> currently work ok.
> 
> I have some ideas than may work with the current release which I will
> try later today, which may solve the advert issue.
> 
> Re mail.ru - does this work OK without indent?

Here I have no problem with mail.ru, but it seems more fast yes
And right now https://americanexpress.com.au/ seems down 

With SSLMITM=off I have no http request at all in Squid 

> 
> What is the symptom when indent is on? firebug trace?

There are massive 407 requests and slowdown (but maybe only related with buggy js)

**Comment by fredbcode** [comment]
> But you are right to be sure I will retry with 3.5.1

Ok same problem with 3.5.1, otherwise it confirms bug #212 there was no 407 in e2 logs with loglevel 2 or 3

**Comment by fredbcode** [comment]
@kenny690  Please can you make a test without identification (removed from Squid and E2) and tell us if you see something new with mail.ru

**Comment by fredbcode** [comment]
> Ok same problem with 3.5.1, otherwise it confirms bug #212 there was no 407 in e2 logs with loglevel 2 or 3

I mean, it's why I don't see this before ...

**Comment by fredbcode** [comment]
Ok with a configuration from kenny I can reproduce random segfault and generate a core.
 
e2guardian[1831]: segfault at 478d178 ip 00007f048e2ffef3 sp 00007f02ecde4238 error 4 in libc-2.24.so[7f048e27c000+195000]

I will prepare a binary with -g for further investigation

**Comment by fredbcode** [comment]
Philip made some changes, please can you make a test ? 
With the new branch : retryfix 

I'm still tracking some segfault

**Comment by Anthony-76** [comment]
Hi, 
I just downloaded the V4.1 release (from source) on Debian 7

I have the same issue,

When I exceutre e2guardian -N, I have the followin error:

SSL_read failed with error 1
[....]
SSL_read failed with error 1
*** Error in `e2guardian': double free or corruption (fasttop): 0x09471e58 ***


e2guardian -v :

e2guardian 4.1.0

Built with:  '--mandir=/usr/share/man/' '--enable-clamd=yes' '--with-proxyuser=dansguardian' '--with-proxygroup=dansguardian' '--prefix=/usr' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--enable-icap=yes' '--enable-commandline=yes' '--enable-trickledm=yes' '--enable-email=yes' '--enable-ntlm=yes' 'CXX=g++' 'CXXFLAGS=-g -O2' 'LDFLAGS=' 'CPPFLAGS=' 'CC=cc' 'CFLAGS=-g -O2' '--enable-sslmitm=yes'

I use the following configuration about ssl:

enablessl = on
cacertificatepath = '/home/e2guardian/my_rootCA.crt'
caprivatekeypath = '/home/e2guardian/private_root.pem'
certprivatekeypath = '/home/e2guardian/private_cert.pem'
sslcertificatepath = '/etc/ssl/certs/'
generatedcertpath = '/home/e2guardian/generatedcerts'


The segfault come from with this site when I click on some links

http://www.msn.com/

Thanks for your help

**Comment by philipianpearce** [comment]
'SSL_read failed with error 1' 

This indicates a general non-specific failure in the openSSL library. Check that you have the most up-to-date version of openssl available on both your development and production platforms. And the same version on both. 

Also, check the info in https://github.com/e2guardian/e2guardian/blob/retryfix/notes/certificate_errors 

Regards 

Philip 

----- Original Message -----

**Comment by Anthony-76** [comment]
Hi,

Thanks for your quick answer.
I updateted the openssl package and the 'SSL_read failed' has disappeared.

Howerover, I have always the : 

*** Error in `e2guardian': double free or corruption (fasttop): 0x08cb22b8 ***
Aborted

**Comment by fredbcode** [comment]
Please, can you make a try with banch "retryfix" ? https://github.com/e2guardian/e2guardian/wiki/Installation-From-Source (selecting source)

**Comment by Anthony-76** [comment]
I just test with the "retryfix".
I have still the same error

*** Error in `e2guardian': double free or corruption (fasttop): 0xb6ed20d8 ***
Aborted

Or
e2guardian -N
SSL_read failed with error 1
Segmentation fault

Theses errors come frm randomly but qucikly about 1-2 minutes after the run.

**Comment by fredbcode** [comment]
Can you try with Debian 8 and exactly

**Comment by fredbcode** [comment]
An exactly the same config, sorry

**Comment by Anthony-76** [comment]
Hi,

It seems better.
I have no segfault for the moment...

**Comment by fredbcode** [comment]
Ok I guess wheezy is too old for v4.x branch, I will put some words

**Comment by fredbcode** [comment]
@Kenny690 there are a lot of changes now, can you make a test with retryfix, please ?

**Comment by Anthony-76** [comment]
After somes tests the problem is still here.

It seems appear when I navigate on the same time o n multi https websites.

**Comment by fredbcode** [comment]
Debian 8 and retryfix ? Please post the result of e2guardian -v and uname -a 
There is a way to reproduce ?

**Comment by Anthony-76** [comment]
Exactly, Debian 8 64bits with retryfix.

in syslog I have this:

May 19 08:57:26 LGL-S e2guardian[2052]: Started sucessfully.
May 19 08:57:37 LGL-S e2guardian[2052]: 51671 No header recd from client at 192.168.170.79 - errno: 0
May 19 08:57:37 LGL-S e2guardian[2052]: 51658 No header recd from client at 192.168.170.79 - errno: 0
May 19 08:57:37 LGL-S e2guardian[2052]: 51649 No header recd from client at 192.168.170.79 - errno: 0
(......)
May 19 08:57:53 LGL-S e2guardian[2052]: 51819 No header recd from client at 192.168.170.79 - errno: 0
May 19 08:57:54 LGL-S kernel: [49786.750504] e2guardian[2095]: segfault at 0 ip 00007f2120c68e0b sp 00007f20ccfe1ec0 error 4 in libc-2.19.so[7f2120bff000+1a1000]



e2guardian -v :-1: e2guardian 4.1.1

Built with:  '--prefix=/usr' '--enable-clamd=yes' '--with-proxyuser=e2guardian' '--with-proxygroup=e2guardian' '--sysconfdir=/etc' '--localstatedir=/var' '--enable-icap=yes' '--enable-commandline=yes' '--enable-email=yes' '--enable-ntlm=yes' '--enable-trickledm=yes' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--enable-pcre=yes' '--enable-sslmitm=yes'

uname -a: 

Linux LGL-S 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2 (2017-04-30) x86_64 GNU/Linux

**Comment by fredbcode** [comment]
Weird, I'm trying exactly with the same OS, can you remove the content of generatedcertpath, compile with '--with-dgdebug=on' and run e2guardian like this 'e2guardian -N' 

Tell us the latest lines before the segfault
Do you think that there is a problem with a specific website ?
Thanks

**Comment by Anthony-76** [comment]
I don't think the problem is link with specific website:

The log before segfault:

Still to read (SSL) 0 bytes
bufferReadFromSocket returned: 17216
Leaving default download manager plugin
52649 -got body from proxy Line: 4263 Function: contentFilter
52649 -Running scanMemory Line: 4309 Function: contentFilter
About to writeMemoryTempFile /tmp/tfO8LAs3 size: 17216
Running: /usr/bin/clamdscan /tmp/tfO8LAs3
Erreur de segmentation

**Comment by fredbcode** [comment]
Ha !

clamd works well for you, you see some files blocked ? Can you make a try without ?

**Comment by Anthony-76** [comment]
I try without, and there is no segfault now.

In e2guardian.conf, the next line:

contentscanner = '/etc/e2guardian/contentscanners/clamdscan.conf'

Doesn't work. When I download a virus via eicar.org, E2guardian don't seems send the file to clamav.

When I activate the next line in e2guardian.conf

contentscanner = '/etc/e2guardian/contentscanners/commandlinescan.conf'

It's work, the virus is blocked but I have the segfault...

**Comment by fredbcode** [comment]
Ok I guess segfault from SSLMITM is fixed now, but there is a problem with AV (seems all plugins ...)

**Comment by fredbcode** [comment]
About commandlinescan perhaps it's an issue with fork and thread a quicker fix should be just remove fork part. 
Just execute execv from thread 

I tried with just /bin/ls it works here 

progname = /bin/ls
virusregexp = (a*)

Can you make the same test ?

Also, do you have some debug returns with clamd ?
You can catch result with e2guardian -N > /tmp/debug

**Comment by fredbcode** [comment]
Maybe you are using some other options like mail in e2guardian.conf ?
Here I have just commandline with /bin/ls as AV

**Comment by fredbcode** [comment]
I'm trying with a more realistic test 

# Program to run & initial arguments - filename for scanning will be appended
progname = /bin/grep yahoo

# At least one of the following three options must be defined!
# They are checked in the following order, with the first match determining
# the scan result:
# virusregexp - regular expression for extracting virus names from
#               the scanner's output
# cleancodes - program return code(s), as a comma-separated list, for
#              uninfected files
# infectedcodes - program return code(s), as a comma-separated list, for
#                 infected files

# virusregexp = (a*)

# Which submatch of the above contains the virus name? (0 = all matched text)
# submatch = 1

cleancodes = 256
infectedcodes = 0,2,3

Any page with "yahoo" is blocked as infected

**Comment by Anthony-76** [comment]
I tried with commandline and progame=/bin/ls.

I have no segfault.

In E2guardian.conf file, I have no option such as mail .

Here the last 100 lines in debug file before segfault:


tail -n 100 /tmp/debug
mime type: image/png Line: 222 Function: isContentType
mimes result : false Line: 249 Function: isContentType
mime type: image/png Line: 222 Function: isContentType
mimes result : false Line: 249 Function: isContentType
55430 -sending header to client Line: 2974 Function: handleConnection
55430  got past line 2677 rfo  Line: 2980 Function: handleConnection
headerout was:HTTP/1.1 200 OK
 Line: 1925 Function: out
headerout is:HTTP/1.1 200 OK
 Line: 1939 Function: out
headertoclient:HTTP/1.1 200 OK
 Line: 1964 Function: out
timeout:20000 Line: 1965 Function: out
 Line: 1981 Function: out19 May 2017 09:33:33 GMT
 Line: 1981 Function: oute: image/png
 Line: 1981 Function: outgth: 28853
 Line: 1981 Function: out keep-alive
 Line: 1981 Function: outn, 14 May 2018 09:33:33 GMT
 Line: 1981 Function: outol: public, max-age=31104000
 Line: 1981 Function: outed: Fri, 28 Apr 2017 09:49:31 GMT
 Line: 1981 Function: outatus: HIT
 Line: 1981 Function: outt-Encoding
 Line: 1981 Function: outes: bytes
 Line: 1981 Function: outudflare-nginx
 Line: 1981 Function: out60e2aec170914-CDG
Header written - pstdata_len:0 Line: 2025 Function: out
Returning from header:out  Line: 2048 Function: out
Client: START------------------------------- Line: 1011 Function: dbshowheader
 Line: 1018 Function: dbshowheaderheader bool: HTTP/1.1 200 OK
 Line: 1018 Function: dbshowheaderheader bool: Date: Fri, 19 May 2017 09:33:33 GMT
 Line: 1018 Function: dbshowheaderheader bool: Content-Type: image/png
 Line: 1018 Function: dbshowheaderheader bool: Content-Length: 28853
 Line: 1018 Function: dbshowheaderheader bool: Connection: keep-alive
 Line: 1018 Function: dbshowheaderheader bool: Expires: Mon, 14 May 2018 09:33:33 GMT
 Line: 1018 Function: dbshowheaderheader bool: Cache-Control: public, max-age=31104000
 Line: 1018 Function: dbshowheaderheader bool: Last-Modified: Fri, 28 Apr 2017 09:49:31 GMT
 Line: 1018 Function: dbshowheaderheader bool: CF-Cache-Status: HIT
 Line: 1018 Function: dbshowheaderheader bool: Vary: Accept-Encoding
 Line: 1018 Function: dbshowheaderheader bool: Accept-Ranges: bytes
 Line: 1018 Function: dbshowheaderheader bool: Server: cloudflare-nginx
 Line: 1018 Function: dbshowheaderheader bool: CF-RAY: 36160e2aec170914-CDG
Client: END------------------------------- Line: 1023 Function: dbshowheader
55430 -sent all header to client Line: 2992 Function: handleConnection
55430 -waschecked:0 Line: 2993 Function: handleConnection
55430 -2tunnel activated Line: 3103 Function: handleConnection
Tunnelling with content length 28853
Data in fdfrom's buffer; sending 991 bytes
Data in fdfrom's buffer; sent 991 bytes
Start of tunnel loop: throughput:991 target:28853
Socket::readFromSocket(ssl): now entering loop to read: 27862 bytes:
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 26493 bytes
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 8752 bytes
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 7383 bytes
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 0 bytes
bufferReadFromSocket returned: 32768
newsize: 32768
dm-default.in: before bcheckForInput:
dm-default.in: after bcheckForInput:
Socket::readFromSocket(ssl): now entering loop to read: 30180 bytes:
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 14267 bytes
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 15784 bytes
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 14415 bytes
Socket::readFromSocket(ssl): before SSL_read:
Still to read (SSL) 0 bytes
bufferReadFromSocket returned: 30180
Leaving default download manager plugin
55457 -got body from proxy Line: 4263 Function: contentFilter
55457 -Running scanMemory Line: 4309 Function: contentFilter
About to writeMemoryTempFile /tmp/tfLw9p9x size: 62948
Scanner result
--------------
/tmp/tfnbvtJm: OK

----------- SCAN SUMMARY -----------
Infected files: 0
Time: 0.020 sec (0 m 0 s)

--------------
Code: 0
Scanner result
--------------
/tmp/tfqgOzos: OK

----------- SCAN SUMMARY -----------
Infected files: 0
Time: 0.007 sec (0 m 0 s)

--------------
Code: 0
55506 -AV scan 0 returned: 0 Line: 4314 Function: contentFilter
55506 -finished running AV Line: 4359 Function: contentFilter
55506 -finished running AV Line: 4359 Function: contentFilter
Running: /usr/bin/clamdscan /tmp/tfLw9p9x

**Comment by fredbcode** [comment]
Can you post your 
contentscanner = '/etc/e2guardian/contentscanners/commandlinescan.conf'
and 
contentscanner = '/etc/e2guardian/contentscanners/clamdscan.conf'

**Comment by Anthony-76** [comment]
/etc/e2guardian/contentscanners/commandlinescan.conf: 

plugname = 'commandlinescan'

exceptionvirusmimetypelist = '/etc/e2guardian/lists/contentscanners/exceptionvirusmimetypelist'
exceptionvirusextensionlist = '/etc/e2guardian/lists/contentscanners/exceptionvirusextensionlist'
exceptionvirussitelist = '/etc/e2guardian/lists/contentscanners/exceptionvirussitelist'
exceptionvirusurllist = '/etc/e2guardian/lists/contentscanners/exceptionvirusurllist'
progname = '/usr/bin/clamdscan'
cleancodes = 0
virusregexp = : ([ -/a-zA-Z0-9\.]+) FOUND
submatch = 1
defaultresult = infected


/etc/e2guardian/contentscanners/clamdscan.conf

plugname = 'clamdscan'

clamdudsfile = '/var/run/clamav/clamd.ctl'
exceptionvirusmimetypelist = '/etc/e2guardian/lists/contentscanners/exceptionvirusmimetypelist'
exceptionvirusextensionlist = '/etc/e2guardian/lists/contentscanners/exceptionvirusextensionlist'
exceptionvirussitelist = '/etc/e2guardian/lists/contentscanners/exceptionvirussitelist'
exceptionvirusurllist = '/etc/e2guardian/lists/contentscanners/exceptionvirusurllist'

**Comment by fredbcode** [comment]
Should be progname = '/usr/bin/clamscan' without "d" ?
Do you have clamscan conf to share, I'm trying here but it's very very slow ...

/tmp/test: OK

----------- SCAN SUMMARY -----------
Known viruses: 6287637
Engine version: 0.99.2
Scanned directories: 0
Scanned files: 1
Infected files: 0
Data scanned: 1.18 MB
Data read: 0.59 MB (ratio 2.01:1)
Time: 25.528 sec (0 m 25 s)

25 for 600ko ...

**Comment by Anthony-76** [comment]
If I use clamscan instead of clamdscan, webpages are veryyyyyyy slow....

clamscan command load the all database at each run

**Comment by fredbcode** [comment]
Ok our discussion is not at the right place, I guess this bug is fixed 
Please take a look at #224

**Comment by marcelloc** [comment]
On BSD system too. Trying with https://americanexpress.com.au/ site

Line: 1018 Function: dbshowheader
Client: END------------------------------- Line: 1023 Function: dbshowheader
63610 got past line 2350 proxy header out Line: 2561 Function: handleConnection
63610 exchange_timeout is 20000 Line: 2562 Function: handleConnection
tunnel tw poll returned ok:1
tunnel got data from sockfrom: 178 bytes
tunnel wrote data out: 178 bytes
Start of tunnel loop: throughput:178 target:-1
tunnel tw poll returned ok:1
Start of tunnel loop: throughput:178 target:-1
getLine !SSL read into buffer; bufflen: 0
getLine terminate string !SSL: 0
firstime: header:in after getLine Line: 2098 Function: in
firstime: header:in after getLine: rc: 0 truncated: 1 Line: 2103 Function: in
17828745671468978451Call : from HTTPHeader.cpp to dbshowheader but header is empty Line: 1028 Function: dbshowheader

**Comment by fredbcode** [comment]
With retryfix and sslmitm ?, strange it worked here

**Comment by fredbcode** [comment]
Ok I can reproduce 
This website **is down now** I think the issue is related 
With an nonexistent domain no problem

wget -v https://americanexpress.com.au
--2017-06-08 10:34:57--  https://americanexpress.com.au/
Résolution de americanexpress.com.au (americanexpress.com.au)… 148.173.91.85
Connexion à americanexpress.com.au (americanexpress.com.au)|148.173.91.85|:443… échec : Connexion terminée par expiration du délai d'attente. **-> Mean timeout**

worker Q size:0
busychildren:1
worker Q size:0
getLine !SSL read into buffer; bufflen: 0
getLine terminate string !SSL: 0
firstime: header:in after getLine  Line: 2098 Function: in
firstime: header:in after getLine: rc: 0 truncated: 1 Line: 2103 Function: in
9249997488325516119Call : from HTTPHeader.cpp to dbshowheader but header is empty Line: 1028 Function: dbshowheader

      
(gdb) bt
#0  0x00007fb7473dd414 in std::string::find(char const*, unsigned long, unsigned long) const () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000000000040926d in String::after(char const*) const ()
#2  0x0000000000433372 in HTTPHeader::returnCode() ()
#3  0x0000000000414c19 in ConnectionHandler::handleConnection(Socket&, String&, bool, Socket&, stat_rec*&) ()
#4  0x000000000040ce7b in ConnectionHandler::handlePeer(Socket&, String&, stat_rec*&) ()
#5  0x0000000000456bb6 in handle_connections(int) ()
#6  0x0000000000466fca in void std::_Bind_simple<void (*(int))(int)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) ()
#7  0x0000000000466e49 in std::_Bind_simple<void (*(int))(int)>::operator()() ()
#8  0x0000000000466d44 in std::thread::_Impl<std::_Bind_simple<void (*(int))(int)> >::_M_run() ()
#9  0x00007fb7473d5970 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#10 0x00007fb747eaa064 in start_thread (arg=0x7fb745019700) at pthread_create.c:309
#11 0x00007fb746b4562d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111

**Comment by fredbcode** [comment]
@marcelloc Please can you make a try, with the latest changes ?

**Comment by marcelloc** [comment]
The first time I tried, the page opened, but few time latter(maybe the same buggy url got offline again) Segmentation fault 

getLine !SSL read into buffer; bufflen: 0
getLine terminate string !SSL: 0
firstime: header:in after getLine  Line: 2104 Function: in
firstime: header:in after getLine: rc: 0 truncated: 1 Line: 2109 Function: in
3814705021384065630Call : from HTTPHeader.cpp to dbshowheader but header is empty Line: 1028 Function: dbshowheader

**Comment by fredbcode** [comment]
Can you try to get a core ? Just to compare with my result above

Le 9 juin 2017 14:02:36 GMT+02:00, Marcello Coutinho <notifications@github.com> a écrit :
>The first time I tried, the page opened, but few time latter(maybe the
>same buggy url got offline again) Segmentation fault 
>
>getLine !SSL read into buffer; bufflen: 0
>getLine terminate string !SSL: 0
>firstime: header:in after getLine  Line: 2104 Function: in
>firstime: header:in after getLine: rc: 0 truncated: 1 Line: 2109
>Function: in
>3814705021384065630Call : from HTTPHeader.cpp to dbshowheader but
>header is empty Line: 1028 Function: dbshowheader
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-307370785

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by marcelloc** [comment]
you mean compile with segfault trace?

**Comment by fredbcode** [comment]
There is some words about in wiki (about Debian) it's a system feature.
After that get the results of:

gdb eguardian core
bt

I will try to reproduce again


Le 10 juin 2017 16:33:15 GMT+02:00, Marcello Coutinho <notifications@github.com> a écrit :
>you mean compile with segfault trace?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/213#issuecomment-307568732

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by marcelloc** [comment]
Does this part of the debug means a non SSL connection?

getLine !SSL read into buffer; bufflen: 0
getLine terminate string !SSL: 0

I'm testing e2g without interception.

**Comment by marcelloc** [comment]
Did more tests tonight. What I found is that when you try any offiline site or ip using http, you get proxyconnect and 504 gateway timed out

pport No header recd from client - errno: 0 Line: 531 Function: handleConnection
55984 -Attempting graceful connection close Line: 3173 Function: handleConnection
handle_peer returned: 0

But if you try the same non responding ip with https, you get segmentation fault. The difference I can see is that https connections is not starting with proxycon but with basesock

tests without ssl mitm

**Comment by fredbcode** [comment]
Yes I agree, it's what I found 
I guess there is something wrong with a string function and a not responding HTTPS website (see my backtrack above).

I will investigate next week, perhaps we can easily reproduce with a kill -9 to squid process

**Comment by fredbcode** [comment]
Marcello, are you sure that you are using the latest changes in "retryfix" ?

git pull ?

**Comment by marcelloc** [comment]
Yes. Got the changes you merged and also the new config file on fx.conf.

**Comment by fredbcode** [comment]
Today the website still down but I can't reproduce, very weird

**Comment by marcelloc** [comment]
Ok. Just try https on any non web server like https://10.10.10.10

**Comment by fredbcode** [comment]
Yes but no
The issue is gone here ...

**Comment by marcelloc** [comment]
I did an workaround here by setting squid to splice all ssl traffic. This way it 'handles' the ssl connection instead of forwarding it directly to the offline host.

**Comment by fredbcode** [comment]
The website should be denied for you ? And are you using BYPASS and bannedsitelistwithbypass ?

**Comment by marcelloc** [comment]
Not using bypass. with this 'semi-interception' e2guardians blocks the page complaining about missing or misconfigured certificate squid delivers with spliceall feature.

This does not fix or identify the issue, but keeps e2guardian alive :)

**Comment by fredbcode** [comment]
marcelloc can you make a try again, I pushed some changes not directly related but they can help

**Comment by marcelloc** [comment]
https issues seams to be gone :D GREAT! 

I got another issue due an custom error page that crashed with this message:
end log request ----------------------------------------
from header url:http://vitrines.globo.com/vitrine/vitrine.min.js Line: 1331 Function: getUrl
Displaying TEMPLATE
-HOST- placeholder encountered but hostname currently unknown; lookup forced.
Segmentation fault

but since I back to normal report, no crashes.

**Comment by marcelloc** [comment]
got some crashes on my production server. I'll push the debug version to see what happened

**Comment by fredbcode** [comment]
You mean offline website, or this issue is gone ?

**Comment by marcelloc** [comment]
the offline https site issue is gone but I'm still getting segfault with traffic.
I'll try to isolate on a specific url or steps that crash

**Comment by fredbcode** [comment]
I pushed some fixes from Philip, can you make a try ?

**Comment by marcelloc** [comment]
still getting crashes. But I could not find a way to identify exactly how to repoduce

Not sure if this last lines before segfault are from the same connect. This log is from my production server.

15103550445763629251REQOUT: dbshowheader bool: Cookie: NID=104=qrppbgN0Jcr1WGdsNOMZNqsrDMDb8GMCcec4um98yGHwW1T33wbbf8Fmiy5aj3vcmoLjoAyWbX3l1wCOJfQBem-5xiFR_9QPs6_8-5z_BDNPZRebZ4XkvU5b0HAXFx5TqKmLOplgyaa Line: 1018 Function: dbshowheader-t9-m9FAvZAzS9uDNGXGkNkYKD1BYLHATSBpPFIWdMLh-XIis36G0fTrxGHFAry092Tjq446pNrNtvERQDb_cNUtfffQJQ9bMjG; CONSENT=YES+BR.pt-BR+20160721-10-0; OGPC=5061821-4:
 Line: 1018 Function: dbshowheaderheader bool: Connection: keep-alive
Client: END------------------------------- Line: 1023 Function: dbshowheader
55647  got past line 2350 proxy header out  Line: 2578 Function: handleConnection
55647  exchange_timeout is 20000 Line: 2579 Function: handleConnection
55647 before docheader in 2371:  Line: 2585 Function: handleConnection
header:in before getLine - firsttime : 1 timeout:20000 Line: 2099 Function: in
in getline (SSL): timeout: 20000 size: 32768
55654 -error/timeout on header in from proxy: 1 Line: 2602 Function: handleConnection
55654 -proxy timedout Line: 2606 Function: handleConnection
55654 -Attempting graceful connection close Line: 3173 Function: handleConnection
getLine !SSL read into buffer; bufflen: 0
getLine terminate string !SSL: 0
header:in after getLine - firsttime : 1 Line: 2103 Function: in
header:in after getLine - firsttime : 1 rc: 0 truncated: 1 Line: 2108 Function: in
681920591235133057Call : from HTTPHeader.cpp to dbshowheader but header is empty Line: 1028 Function: dbshowheader
Segmentation fault
